File: raptor.js
"use strict"


class RemoteException extends Error {
    constructor(message) {
        super(message)
        this.name = "RemoteException"
    }
}


class InvalidHTTPStatus extends Error {
    constructor(status) {
        super(status)
        this.name = "InvalidHTTPStatus"
    }
}


class ConnectionError extends Error {
    constructor() {
        super("ConnectionError")
        this.name = "ConnectionError"
    }
}


class InvalidResponse extends Error {
    constructor() {
        super("InvalidResponse")
        this.name = "InvalidResponse"
    }
}


class InvalidResponseObject extends InvalidResponse {
    constructor() {
        super("InvalidResponseObject")
        this.name = "InvalidResponseObject"
    }
}


class InvalidJson extends InvalidResponse {
    constructor() {
        super("InvalidJson")
        this.name = "InvalidJson"
    }
}


class RemoteServer {
    constructor(endpoint, token = "public") {
        this.functions = new Map

        this.endpoint = endpoint
        this.token = token

        this.proxy = new Proxy(this, {

            get(target, prop) {
                if (prop == "functions") {
                    return target.functions
                }

                if (target.functions.get(prop) != undefined) {
                    return target.functions.get(prop)
                }

                console.log("creating")
                let func = function() {
                    let args_array = []
                    for (const arg of arguments) {
                        args_array.push(arg)
                    }

                    let request_object = {
                        token: target.token,
                        function: {
                            name: prop,
                            args: args_array,
                            kwargs: {}
                        }
                    }

                    try {
                        var request = new XMLHttpRequest();
                        request.open('POST', target.endpoint, false);  // `false` makes the request synchronous
                        request.send(JSON.stringify(request_object));
                    } catch {
                        throw new ConnectionError()
                        return undefined
                    }

                    if (request.status !== 200) {
                        throw new InvalidHTTPStatus(request.status)
                        return undefined
                    }

                    let response_object

                    try {
                        response_object = JSON.parse(request.responseText)
                    } catch {
                        throw new InvalidJson()
                        return undefined
                    }

                    let remote_exception, return_object

                    try {
                        remote_exception = response_object["exception"]
                        return_object = response_object["return"]
                    } catch {
                        throw new InvalidResponseObject()
                        return undefined
                    }

                    if (remote_exception !== "") {
                        throw new RemoteException(remote_exception)
                    }

                    if (return_object === "NoneType") {
                        return undefined
                    }

                    return return_object
                    
                }

                target.functions.set(prop, func)
                return func
                
            }

        })

        return this.proxy
    }

}

File: raptor.py
import requests
import json


class RemoteException(Exception):
    pass


class InvalidResponse(Exception):
    pass


class InvalidJson(InvalidResponse):
    pass


class InvalidResponseObject(InvalidResponse):
    pass


class ConnectionError(Exception):
    pass


class InvalidHTTPStatus(Exception):
    pass
        

class RemoteFunction():
    def __init__(self, endpoint: str, remote_function_name: str, token: str) -> None:
        self.remote_function_name = remote_function_name
        self.endpoint = endpoint
        self.token = token

        # создаём объект, из которого будет делаться отправляемый json
        self.object = {}
        self.object["token"] = self.token
        self.object["function"] = {}
        self.object["function"]["name"] = self.remote_function_name

    def __call__(self, *args, **kwargs):
        self.object["function"]["args"] = args
        self.object["function"]["kwargs"] = kwargs
        post_json = json.dumps(self.object)

        try:
            response = requests.post(self.endpoint, data=post_json, headers={"Content-Type": "application/json"})
        except:
            raise ConnectionError
            return

        if response.status_code != 200:
            raise InvalidHTTPStatus

        try:
            response_object = json.loads(response.text)
        except:
            raise InvalidJson
            return

        try:
            remote_exception = response_object["exception"]
            return_object = response_object["return"]
        except:
            raise InvalidResponseObject
            return
        
        if remote_exception != "":
            raise RemoteException(remote_exception)
            return

        if return_object == "NoneType":
            return

        return return_object


class RemoteServer():
    def __init__(self, endpoint: str, token: str = "public") -> None:
        self.endpoint = endpoint
        self.token = token
        self.functions = {}

    def __getattr__(self, __name: str):
        try:
            remote_function = self.functions[__name]
        except KeyError:
            remote_function = RemoteFunction(self.endpoint, __name, self.token)
            self.functions[__name] = remote_function
        
        return remote_function

    def rebind(self, new_endpoint: str, new_token: str):
        self.endpoint = new_endpoint
        self.token = new_token

        for item in self.functions.items():
            item.endpoint = new_endpoint
            item.token = new_token

            

if __name__ == "__main__":
    server = RemoteServer("http://127.0.0.1:5000/", "token")
    azaza = server.func(33, "ffew", lol="33")
    azaza = server.func(355, "ffew", lol="33")
    azaza = server.func(34, "ffff", lol="334")
    print(type(azaza))
    print(azaza)
    print(server.tee(455))

        
